src/osiris_log.erl

```diff
@@ -1884,6 +1821,72 @@ open_index_read(File) ->
     _ = file:read(Fd, ?IDX_HEADER_SIZE),
     Fd.
 
+scan_idx(Offset, SegmentInfo = #seg_info{index = IndexFile, last = LastChunkInSegment}) ->
+    case range_from_segment_infos([SegmentInfo]) of
+        empty -> 
+            %% if the index is empty do we really know the offset will be next
+            %% this relies on us always reducing the Offset to within the log range
+            {0, ?LOG_HEADER_SIZE};
+        {SegmentStart, SegmentEnd} ->
+            case Offset < SegmentStart orelse Offset > SegmentEnd + 1 of
+                true -> offset_out_of_range;
+                false ->
+                    IndexFd = open_index_read(IndexFile),
+                    Result = scan_idx(IndexFd, Offset, LastChunkInSegment),
+                    file:close(IndexFd),
+                    Result
+            end
+    end.
+
+scan_idx(Fd, Offset, #chunk_info{id = LastChunkInSegmentId, num = LastChunkInSegmentNum}) ->
+    case file:read(Fd, ?INDEX_RECORD_SIZE_B) of
+        {ok,
+         <<ChunkId:64/unsigned,
+           _Timestamp:64/signed,
+           _Epoch:64/unsigned,
+           FilePos:32/unsigned>>} ->
+            case Offset < ChunkId of
+                true ->
+                    %% offset is lower than the first chunk in this segment
+                    %% shouldn't really happen as we check the range above
+                    offset_out_of_range;
+                false ->
+                    case Offset > (LastChunkInSegmentId + LastChunkInSegmentNum - 1) of
+                        true ->
+                            %% offset higher than the last chunk in this segment
+                            {LastChunkInSegmentId + LastChunkInSegmentNum, eof};
+                        false ->
+                            %% offset is in this segment
+                            scan_idx(Fd, Offset, {ChunkId, FilePos})
+                    end
+            end;
+        eof ->
+            %% this should never happen - offset is in the range and we are raeding the first record
+            {error, eof};
```

Comment:
the calling code relies on this being just `eof` not `{error, eof}` - surprised dialyzer didn't catch that.